lib/class-wp-theme-json-gutenberg.php

```diff
@@ -1897,9 +1941,30 @@ protected static function compute_style_properties( $styles, $settings = array()
 			return $declarations;
 		}
 
-		$root_variable_duplicates = array();
+		/*
+		 * Instead of iterating over the $property list (54 items as of this comment),
+		 * we iterate only over the properties present in the $styles array.
+		 *
+		 * To gauge the impact this change has, let's take a look at the numbers for
+		 * the TwentyTwentyThree theme at the time of this comment.
+		 *
+		 * - The compute_style_properties method  is called 65 times, one per style node.
+		 * - The $properties list has 54 items. For TT3, $styles may have less than 5 items per node.
+		 *
+		 * - Before: 65 * 54 = 3510 times.
+		 * - After:  65 *  5 =  325 times.
+		 *
+		 */
+		$paths_in_node         = self::get_paths_in_node( $styles );
+		$properties_to_inspect = array();
+		foreach ( $properties as $key => $value ) {
+			if ( in_array( $value, $paths_in_node, true ) ) {
+				$properties_to_inspect[ $key ] = $value;
+			}
+		}
 
-		foreach ( $properties as $css_property => $value_path ) {
+		$root_variable_duplicates = array();
+		foreach ( $properties_to_inspect as $css_property => $value_path ) {
 			$value = static::get_property_value( $styles, $value_path, $theme_json );
 
 			if ( str_starts_with( $css_property, '--wp--style--root--' ) && ( static::ROOT_BLOCK_SELECTOR !== $selector || ! $use_root_padding ) ) {
```

Comment:
```suggestion
			if ( ( static::ROOT_BLOCK_SELECTOR !== $selector || ! $use_root_padding ) && str_starts_with( $css_property, '--wp--style--root--' )  ) {
```

Changing this line also helps performance.